### PR TITLE
support bond type gt4 in MDLV3000Writer

### PR DESCRIPTION
**Pull request by uli-f**
First draft.

As `Expr` can be arbitrarily complex I decided to only deal with a very simple subset.

Test `MDLV3000WriterTest::writeBondTypeFiveTest` is currently failing when atom valence is inferred as `bond.getOrder()` comes back with `null` for atom that is part of the query bond.

@johnmay I was hoping yo get your opinion on how to best approach this. Do atoms that are part of query bond need to be converted to query atoms to avoid inferring their explicit valence?

### CODE REVIEW

File: storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Writer.java [hunk 1]
```diff
@@ -513,18 +513,22 @@ private IAtom[] pushHydrogensToBack(IAtomContainer mol, Map<IChemObject, Integer
             throw new AssertionError("Map atomToIdx must be empty.");
 
         final IAtom[] atoms = new IAtom[mol.getAtomCount()];
+        final List<IAtom> hydrogenAtoms = new ArrayList<>();
```
Comment (johnmay) [inline storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Writer.java:558]: Why was this needed? It's fine but actually the second copy array is marginally less efficient - but in practice it won't mater. Technically a dutch national flag like algorithm would be best but this function isn't critical.

File: storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Writer.java [hunk 2]
```diff
@@ -445,15 +448,54 @@ private void writeBondBlock(IAtomContainer mol,
...
+                    final Expr expression = ((QueryBond) bond).getExpression();
+                    // Expression are predicate trees and can be arbitrarily complex.
+                    // Only simple expressions consisting of a single node with a select few types indicating
+                    // a bond order are considered here.
+                    if (expression.left() != null || expression.right() != null) {
+                        throw new CDKException("Query bonds whose features are described by more than a single Expressions " +
+                                bond.getOrder() + " cannot be written to V3000");
+                    }
+                    switch (expression.type()) {
+                        case SINGLE_OR_DOUBLE:
```
Comment (johnmay) [inline storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Writer.java:474]: 1) Should also check the OR(SINGLE,DOUBLE) OR(DOUBLE,SINGLE) , OR(SINGLE,AROM) etc. variants as they can technically occur if a user constructs them. 

Technically something daft like OR(DOUBLE,OR(SINGLE,FALSE)) is also equivalent. It would be easier if to a "canonical" expression but we don't have a way to do that. One option is you generate bond code which tells you what you match (let me have a go at writing a small function which does that). 

I also think bondType=8 should be the default in non-strict mode, this is common in other tools. "There is a query

**Comment by johnmay** [comment]
Here is the gist of the canonical bond code/expressions.

Something I forgot to add is there is subtle semantics between SMARTS and MDL, in MDL SINGLE_OR_DOUBLE may match aromatic bonds as well. There is no equivalent in SMARTS since all single/double expres are implicitly aliphatics


```java
    private static final int CHAIN_SING   = 0x0001; // 1 << 0
    private static final int CHAIN_DOUB   = 0x0002; // 1 << 1
    private static final int CHAIN_TRIP   = 0x0004; // 1 << 2
    private static final int CHAIN_QUAD   = 0x0008; // 1 << 3
    private static final int RING_SING    = 0x0010; // 1 << 4
    private static final int RING_DOUB    = 0x0020; // 1 << 5
    private static final int RING_TRIP    = 0x0040; // 1 << 6
    private static final int RING_QUAD    = 0x0080; // 1 << 7
    private static final int AROM_SING    = 0x0100; // 1 << 8
    private static final int AROM_DOUB    = 0x0200; // 1 << 9

    private static final int SING_MASK    = 0x0111;
    private static final int DOUB_MASK    = 0x0222;
    private static final int TRIP_MASK    = 0x0044;
    private static final int QUAD_MASK    = 0x0088;
    private static final int CHAIN_MASK   = 0x000f;
    private static final int RING_MASK    = 0x03f0;
    private static final int AROM_MASK    = 0x0300;
    private static final int ALIP_MASK    = 0x00ff;
    private static final int ALL_MASK     = 0x03ff;

    int getBondCode(Expr expr) {
        switch (expr.type()) {
            case NOT:
                return ~getBondCode(expr.left());
            case OR:
                return getBondCode(expr.left()) | getBondCode(expr.right());
            case AND:
                return getBondCode(expr.left()) & getBondCode(expr.right());
            case TRUE:
            case STEREOCHEMISTRY:
                return ALL_MASK;
            case SINGLE_OR_AROMATIC:
                return SING_MASK | AROM_MASK;
            case DOUBLE_OR_AROMATIC:
                return DOUB_MASK | AROM_MASK;
            case SINGLE_OR_DOUBLE:
                return SING_MASK | DOUB_MASK;
            case IS_AROMATIC:
                return AROM_MASK;
            case IS_ALIPHATIC:
                return ALIP_MASK;
            case IS_IN_RING:
                return RING_MASK;
            case IS_IN_CHAIN:
                return CHAIN_MASK;
            case ORDER:
                switch (expr.value()) {
                    case 1:
                        return SING_MASK;
                    case 2:
                        return DOUB_MASK;
                    case 3:
                        return TRIP_MASK;
                    case 4:
                        return QUAD_MASK;
                    default:
                        return ALL_MASK;
                }
            case ALIPHATIC_ORDER:
                switch (expr.value()) {
                    case 1:
                        return SING_MASK & ALIP_MASK;
                    case 2:
                        return DOUB_MASK & ALIP_MASK;
                    case 3:
                        return TRIP_MASK & ALIP_MASK;
                    case 4:
                        return QUAD_MASK & ALIP_MASK;
                    default:
                        return ALL_MASK;
                }
            case FALSE:
            default:
                return 0;
        }
    }
```

File: storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Writer.java (hunk 1 above)
Comment (uli-f) [reply storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Writer.java]: I apologize, that wasn't the best of changes. I changed the implementation in `MDLV3000Writer::pushHydrogensToBack` in commit [fc06c89](https://github.com/cdk/cdk/pull/1106/commits/fc06c89c4bfa66949e06e06dcf43c4dfe40e351c) to something that I would consider along the lines of a 2 way partitioning algorithm. But happy to go back on that change if you don't like it.

File: storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Writer.java (hunk 2 above)
Comment (uli-f) [reply storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Writer.java]: > I also think bondType=8 should be the default in non-strict mode, this is common in other tools.

I remember seeing `IChemObjectReader.Mode.{Relaxed|Strict}` but I cannot find anything analogous in `IChemObjectWriter`. Not sure, if I understand what you referring to here.

**Comment by uli-f** [comment]
> Here is the gist of the canonical bond code/expressions.

Thanks for pointing me in the right direction here.

**Comment by uli-f** [comment]
@johnmay This brings me back to the original issue feel a bit stuck with: 

> Test `MDLV3000WriterTest::writeBondTypeFiveTest` is currently failing when atom valence is inferred as `bond.getOrder()` comes back with null for an atom that is part of the query bond.

It's pretty obvious what's happening once you run `MDLV3000WriterTest::writeBondTypeFiveTest`.

Would appreciate your input how to best deal with this.

**Comment by johnmay** [comment]
We don't enforce this but you shouldn't really be mixing query atoms/bonds and regular atoms/bonds. That's not how the CDK APIs are intended to be used - it would be nice if it did work like that but you're going to have a lot of problems elsewhere. 

The fix for that test is it should not be setting the valence on an atom which is connected to a query bond.

**Comment by uli-f** [comment]
> We don't enforce this but you shouldn't really be mixing query atoms/bonds and regular atoms/bonds. That's not how the CDK APIs are intended to be used - it would be nice if it did work like that but you're going to have a lot of problems elsewhere.

That makes sense, thanks for pointing it out. So in your opinion what would be a clean(er) way to serialize and deserialize objects with query features? Put query atoms/bonds into query AtomContainers to at least make it easier for algorithms in CDK to warn/reject inputs with query features? Or explicitly make all constituents (atoms, bonds) of an AtomContainer with a query feature query constituents to make it obvious that valences etc cannot be inferred for some of them? Just curious here...

> The fix for that test is it should not be setting the valence on an atom which is connected to a query bond.

Thank you. I did that and can now round-trip V3000 files with query bond types 5-8.

One more thing I'd like to put in is the query feature _topology_ for bonds.

**Comment by uli-f** [comment]
> One more thing I'd like to put in is the query feature _topology_ for bonds.

Done.

@johnmay Looking forward to your feedback :)

File: storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Reader.java [hunk 3]
```diff
@@ -878,6 +878,34 @@ private void readBondBlock(ReadState state) throws CDKException {
...
+                                                final Expr expressionBondTypeAndInRing = bondTypeExpression.and(new Expr(Expr.Type.IS_IN_RING));
+                                                ((QueryBond) bond).setExpression(expressionBondTypeAndInRing);
+                                                break;
+                                            // chain
+                                            case 2:
+                                                final Expr expressionBondTypeAndInChain = bondTypeExpression.and(new Expr(Expr.Type.IS_IN_CHAIN));
+                                                ((QueryBond) bond).setExpression(expressionBondTypeAndInChain);
+                                                break;
+                                            default:
+                                                logger.warn("Invalid value " + topology + " for key " + key);
```
Comment (johnmay) [inline storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Reader.java:902]: handleError() ?

File: storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Reader.java [hunk 4]
```diff
@@ -878,6 +878,34 @@ private void readBondBlock(ReadState state) throws CDKException {
...
+                                                break;
+                                        }
+                                    } else {
+                                        logger.warn("Key " + key + " is only defined for query bonds");
```
Comment (johnmay) [inline storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Reader.java:906]: handleError() ?

File: storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Writer.java [hunk 5]
```diff
@@ -515,13 +528,15 @@ private void writeBondBlock(IAtomContainer mol,
                         throw new CDKException("Bond with bond order " + bond.getOrder() + " that isn't flagged as aromatic cannot be written to V3000");
                     }
                 } else if (bond instanceof IQueryBond) {
+                    // Bond needs to be dereferenced to assess actual implementing class of bond.
+                    final IQueryBond dereferencedBond = (IQueryBond) BondRef.deref(bond);
                     // Only query bonds of the class QueryBond are supported as the actual query expression needs to be
                     // extracted from the bond.
-                    if (!(bond instanceof QueryBond)) {
-                        throw new CDKException("Query bond of type " + bond.getClass() + " cannot be written to V3000");
+                    if (!(dereferencedBond instanceof QueryBond)) {
```
Comment (johnmay) [inline storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Writer.java:535]: Note: The reference bond will still be an IQueryBond.... but this is fine

**Review by johnmay — APPROVED** [review]
All OK but I would probably use handleError() for the warnings. It's OK if you don't know the line number etc

**Comment by uli-f** [comment]
Replaced the two logging calls with calls to `handleError()`.

However, I had a look at the `handleError()` method in `DefaultChemObjectReader` and I couldn't find any implementation of `IChemObjectReaderErrorHandler`. So the error handler looks more like framework concept to me without an actual runtime impact.

Also, now nothing gets logged any more. So might be better to do both, i.e. log a warning and call `handleError()`..?

**Comment by egonw** [comment]
> So the error handler looks more like framework concept to me without an actual runtime impact.

The idea is indeed that users of the CDK library implement the error handling. Bioclipse did that, and passed the messages to the proper Eclipse place. For warnings and errors, this was picked up by the editor so that the errors can be visualized in the editor.

See also: https://chem-bla-ics.blogspot.com/2010/01/validating-mdl-sd-files-and-symyx.html

File: storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Reader.java [hunk 6]
```diff
@@ -899,11 +899,11 @@ private void readBondBlock(ReadState state) throws CDKException {
                                                 ((QueryBond) bond).setExpression(expressionBondTypeAndInChain);
                                                 break;
                                             default:
-                                                logger.warn("Invalid value " + topology + " for key " + key);
+                                                handleError("Invalid value " + topology + " for key " + key);
```
Comment (egonw) [inline storage/ctab/src/main/java/org/openscience/cdk/io/MDLV3000Reader.java:902]: if possible, plz also pass the line number and perhaps even the column number

**Comment by johnmay** [comment]
Good to merge?

**Comment by uli-f** [comment]
> Good to merge?

Yes please.
